rfd/0050-join-methods.md

```diff
@@ -0,0 +1,117 @@
+
+The joining node also needs to be able to trust that the Teleport cluster is
+authentic. It can do this by using CA pins configured on the node, in the case
+of IoT nodes connecting to the proxy, the proxy TLS cert can be trusted through
+PKI.
+
+### Current Join Endpoints
+
+#### Auth `POST /tokens/register`
+Join methods: token, EC2
+Server trust: CA pins
+This is the main token register endpoint on the Auth server. Accepts tokens or
+EC2 Identity Documents to authenticate the client.
+
+#### Auth `POST /tokens/register/auth`
+Join Methods: token
+Server trust: CA pins
+Looks like this is a legacy endpoint the just deletes the given token. Only
+works for tokens with the `Auth` role. I can't find anything that calls this, so
+it can probably be scheduled for deletion.
+
+#### Auth `rpc GenerateHostCerts(HostCertsRequest) returns (Certs)`
+Client trust: mTLS
+Server trust: mTLS
+This is an mTLS authenticated gRPC endpoint for cert renewal. The node must
+first join the cluster using another endpoint to get its first client
+certificate used for the mTLS connection.
+
+#### Proxy `POST /webapi/host/credentials`
+Join methods: token, EC2
+Server trust: PKI
+This is the proxy endpoint for registering IoT nodes that don't have a direct connection to the auth server, it basically forwards to
+the Auth `/tokens/register` endpoint.
+
+### IAM Join Method
+The IAM join method (introduced in RFD 41) requires gRPC methods rather than
+HTTP because the design requires streams to implement a challenge/response
+protocol. To summarize, after the client initiates the rpc the Auth server sends
+a crypto-random "challenge" string over the server->client stream. This
+challenge must be included in a signed `sts:GetCallerIdentity` request which
+will be sent on the client->server stream.  Finally, if everything checks out,
+the Auth server will send the signed certificates on the server->client stream.
+
+A new gRPC streaming method will be added on Auth and Proxy to support this. The
+client will be able to call the gRPC method with either a Proxy or Auth address
+and it will "just work" transparently.
+
+#### Auth `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join mathods: IAM
+Server trust: CA pins
+This is the Auth endpoint that will complete the IAM join request.
+
+#### Proxy `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join mathods: IAM
+Server trust: PKI
+Normally, authenticated gRPC calls from IoT nodes are tunnelled through the
+proxy over SSH. This will not work for unauthenticated clients which don't yet
+have an SSH certificate.
+
+It would be possible to use TLS routing on the proxy to forward gRPC requests
+directly to the Auth server, but this would open up the Auth server to DOS
+attacks from unauthenticated clients.
+
+Instead, this single gRPC method will be implemented on the proxy. This method
```

Comment:
We have this same chicken-and-egg problem for would-be authentication RPCs too. I had considered before having a non-mTLS Auth gRPC service for such methods, and this sounds like a good fit for that (hypothetical) service as well.

As an implementation detail we could make it so that the exact same gRPC service is implemented by both Proxy and Auth - Proxy allows non-mTLS access and calls Auth, Auth requires mTLS, achieving the very same scenario you describe.

(Incidentally, this is also a good step in moving us to multiple, small RPC services instead of a single gigantic one.)

Reply:
Yeah I agree it should be the exact same gRPC service. But I don't think Auth can require mTLS, non-IoT nodes need to call Auth directly in the current implementation. Maybe we could move to a world where non-IoT nodes always hit the Proxy eventually, there would be some benefits like leveraging PKI instead of CA pins and being able to further lock down Auth.